lib/Post.php

```diff
@@ -574,89 +574,167 @@ public function comment_form( $args = array() ) {
+	 *     @type string $term_class The Timber term class to use for the term objects.
+	 * }
+	 * @param bool   $merge      Deprecated. Optional. See `$merge` argument in `$args` parameter.
+	 * @param string $term_class Deprecated. Optional. See `$term_class` argument in `$args`
+	 *                           parameter.
+	 * @return array An array of taxonomies.
+	 */
+	public function terms( $args = array(), $merge = true, $term_class = '' ) {
+		// Ensure backwards compatibility.
+		if ( ! is_array( $args ) || isset( $args[0] ) ) {
+			$args = array(
+				'query' => array(
+					'taxonomy' => $args,
+				),
+				'merge' => $merge,
+				'term_class' => $term_class,
+			);
+
+			if ( empty( $args['term_class']) ) {
+				$args['term_class'] = $this->TermClass;
+			}
+		}
+
+		// Defaults.
+		$args = wp_parse_args( $args, array(
+			'query' => array(
+				'taxonomy' => 'all',
+			),
+			'merge' => true,
+			'term_class' => $this->TermClass,
+		) );
+
+		$tax        = $args['query']['taxonomy'];
+		$merge      = $args['merge'];
+		$term_class = $args['term_class'];
+
 		$taxonomies = array();
-		$TermClass = $TermClass ?: $this->TermClass;
 
+		// @todo: Remove in 2.x
 		if ( is_string($merge) && class_exists($merge) ) {
-			$TermClass = $merge;
+			$term_class = $merge;
 		}
-		if ( is_array($tax) ) {
+
+		// Build an array of taxonomies.
+		if ( is_array( $tax ) ) {
 			$taxonomies = $tax;
-		}
-		if ( is_string($tax) ) {
-			if ( in_array($tax, array('all', 'any', '')) ) {
+		} elseif ( is_string( $tax ) ) {
+			if ( in_array( $tax, array( 'all', 'any', '' ) ) ) {
 				$taxonomies = get_object_taxonomies($this->post_type);
 			} else {
 				$taxonomies = array($tax);
 			}
 		}
 
-		$term_class_objects = array();
-
-		foreach ( $taxonomies as $taxonomy ) {
-			if ( in_array($taxonomy, array('tag', 'tags')) ) {
+		// @todo Remove in 2.x
+		$taxonomies = array_map( function( $taxonomy ) {
+			if ( in_array( $taxonomy, array( 'tag', 'tags' ), true ) ) {
 				$taxonomy = 'post_tag';
-			}
-			if ( $taxonomy == 'categories' ) {
+			} elseif ( 'categories' === $taxonomy ) {
 				$taxonomy = 'category';
 			}
 
-			$terms = wp_get_post_terms($this->ID, $taxonomy);
+			return $taxonomy;
+		}, $taxonomies );
 
-			if ( is_wp_error($terms) ) {
-				/* @var $terms WP_Error */
-				Helper::error_log("Error retrieving terms for taxonomy '$taxonomy' on a post in timber-post.php");
-				Helper::error_log('tax = '.print_r($tax, true));
-				Helper::error_log('WP_Error: '.$terms->get_error_message());
+		$terms = wp_get_post_terms( $this->ID, $taxonomies, $args['query'] );
 
-				return $term_class_objects;
-			}
+		if ( is_wp_error( $terms ) ) {
+			/**
+			 * @var $terms \WP_Error
+			 */
+			Helper::error_log( "Error retrieving terms for taxonomies on a post in timber-post.php" );
+			Helper::error_log( 'tax = ' . print_r( $tax, true ) );
+			Helper::error_log( 'WP_Error: ' . $terms->get_error_message() );
 
```

Comment:
Currently `TestTimberPostTerms::testTermExceptions()` is failing b/c of this portion of code. Since `$terms` is an error by this point, should we return an empty array or `null` at this point?

Reply:
Nevermind, on further inspection it appears that the test was actually faulty in trying to fetch terms from a non-existent taxonomy